### PR TITLE
Change AuthTokenContext listener to pass UserInformation on user change

### PR DESCRIPTION
**Pull request by jinquan2019**
Please have a look at our [guidelines for contributions](https://github.com/Microsoft/AppCenter-SDK-Android/blob/develop/CONTRIBUTING.md) and consider the following before you submit the PR:

* [ ] Has `CHANGELOG.md` been updated?
* [ ] Are tests passing locally?
* [x] Are the files formatted correctly?
* [ ] Did you add unit tests?
* [ ] Did you test your change with either the sample apps that are included in the repository or with a blank app that uses your change?
* [ ] Did you check UI tests on the sample app? They are not executed on CI.

## Description

Pass user information instance instead of auth token on user change.

### CODE REVIEW

File: sdk/appcenter/src/test/java/com/microsoft/appcenter/utils/context/AuthTokenContextTest.java [hunk 1]
```diff
@@ -109,7 +109,7 @@ public void setAuthTokenTest() {
 
         /* Verify that listener is called. */
         verify(mockListener, times(2)).onNewAuthToken(notNull(String.class));
-        verify(mockListener, times(1)).onNewUser(notNull(String.class));
+        verify(mockListener, times(1)).onNewUser(notNull(UserInformation.class));
```
Comment (guperrot) [inline sdk/appcenter/src/test/java/com/microsoft/appcenter/utils/context/AuthTokenContextTest.java]: Not a good test, we need to capture userInfo and check accountId

File: sdk/appcenter/src/main/java/com/microsoft/appcenter/utils/context/AuthTokenContext.java [hunk 2]
```diff
@@ -163,7 +163,8 @@ public void setAuthToken(String authToken, String homeAccountId, Date expiresOn)
         for (Listener listener : mListeners) {
             listener.onNewAuthToken(authToken);
             if (isNewUser) {
-                listener.onNewUser(authToken);
+                UserInformation userInfo = homeAccountId == null ? null : new UserInformation(homeAccountId);
```
[the remark this answers is not in the document]
Comment (guperrot) [reply sdk/appcenter/src/main/java/com/microsoft/appcenter/utils/context/AuthTokenContext.java]: You to string.split to keep only the guid part since we store only the long version I think

File: sdk/appcenter/src/main/java/com/microsoft/appcenter/utils/context/AuthTokenContext.java [hunk 3]
```diff
@@ -46,6 +46,12 @@
     @VisibleForTesting
     static final int TOKEN_HISTORY_LIMIT = 5;
 
+    /**
+     * The minimum length of account Id.
```
Comment (jwallra) [inline sdk/appcenter/src/main/java/com/microsoft/appcenter/utils/context/AuthTokenContext.java]: This is maximum length because when we use it on line 172 we're using it as a parameter for min, which means we take the shorter of the two substring lengths.

```suggestion
     * The Maximum length of account Id.
```

File: sdk/appcenter/src/main/java/com/microsoft/appcenter/utils/context/AuthTokenContext.java [hunk 4]
```diff
@@ -46,6 +46,12 @@
...
+     */
+    @VisibleForTesting
+    static final int ACCOUNT_ID_LENGTH = 36;
```
Comment (jwallra) [inline sdk/appcenter/src/main/java/com/microsoft/appcenter/utils/context/AuthTokenContext.java]: I would suggest adding 'max' to the variable name, e.g. `ACCOUNT_ID_MAX_LENGTH`

File: sdk/appcenter/src/test/java/com/microsoft/appcenter/utils/context/AuthTokenContextTest.java [hunk 5]
```diff
@@ -109,7 +110,9 @@ public void setAuthTokenTest() {
 
         /* Verify that listener is called. */
         verify(mockListener, times(2)).onNewAuthToken(notNull(String.class));
-        verify(mockListener, times(1)).onNewUser(notNull(String.class));
+        ArgumentCaptor<UserInformation> captorArg = ArgumentCaptor.forClass(UserInformation.class);
+        verify(mockListener, times(1)).onNewUser(captorArg.capture());
```
Comment (jwallra) [inline sdk/appcenter/src/test/java/com/microsoft/appcenter/utils/context/AuthTokenContextTest.java]: `times(1)` is always assumed unless specified.

```suggestion
        verify(mockListener).onNewUser(captorArg.capture());
```

File: sdk/appcenter/src/test/java/com/microsoft/appcenter/utils/context/AuthTokenContextTest.java [hunk 6]
```diff
@@ -119,7 +122,7 @@ public void setAuthTokenTest() {
 
         /* Verify that listener is called on empty token. */
         verify(mockListener, times(1)).onNewAuthToken(isNull(String.class));
-        verify(mockListener, times(1)).onNewUser(isNull(String.class));
+        verify(mockListener, times(1)).onNewUser(isNull(UserInformation.class));
```
Comment (jwallra) [inline sdk/appcenter/src/test/java/com/microsoft/appcenter/utils/context/AuthTokenContextTest.java]:
```suggestion
        verify(mockListener).onNewUser(isNull(UserInformation.class));
```

File: sdk/appcenter/src/test/java/com/microsoft/appcenter/utils/context/AuthTokenContextTest.java [hunk 7]
```diff
@@ -119,7 +122,7 @@ public void setAuthTokenTest() {
 
         /* Verify that listener is called on empty token. */
         verify(mockListener, times(1)).onNewAuthToken(isNull(String.class));
```
Comment (jwallra) [inline sdk/appcenter/src/test/java/com/microsoft/appcenter/utils/context/AuthTokenContextTest.java]:
```suggestion
         verify(mockListener).onNewAuthToken(isNull(String.class));
```

File: sdk/appcenter-identity/src/main/java/com/microsoft/appcenter/identity/Identity.java [hunk 8]
```diff
@@ -29,6 +29,7 @@
 import com.microsoft.appcenter.utils.async.DefaultAppCenterFuture;
 import com.microsoft.appcenter.utils.context.AbstractTokenContextListener;
 import com.microsoft.appcenter.utils.context.AuthTokenContext;
+import com.microsoft.appcenter.utils.context.UserInformation;
```
Comment (guperrot) [inline sdk/appcenter-identity/src/main/java/com/microsoft/appcenter/identity/Identity.java]: Optimize imports, it is suspicious that an import is added and not used.

File: sdk/appcenter-identity/src/main/java/com/microsoft/appcenter/identity/SignInResult.java [hunk 9]
```diff
@@ -5,6 +5,8 @@
 
 package com.microsoft.appcenter.identity;
 
+import com.microsoft.appcenter.utils.context.UserInformation;
```
Comment (guperrot) [inline sdk/appcenter-identity/src/main/java/com/microsoft/appcenter/identity/SignInResult.java]: Optimize imports, it is suspicious that an import is added and not used.

File: sdk/appcenter/src/main/java/com/microsoft/appcenter/utils/context/AuthTokenContext.java (hunk 4 above)
Comment (guperrot) [reply sdk/appcenter/src/main/java/com/microsoft/appcenter/utils/context/AuthTokenContext.java]: It's fixed length, not variable so always max. The code for min is just a guard.

File: sdk/appcenter/src/main/java/com/microsoft/appcenter/utils/context/UserInformation.java [hunk 10]
```diff
@@ -3,7 +3,7 @@
  * Licensed under the MIT License.
  */
 
-package com.microsoft.appcenter.identity;
+package com.microsoft.appcenter.utils.context;
```
Comment (guperrot) [inline sdk/appcenter/src/main/java/com/microsoft/appcenter/utils/context/UserInformation.java]: UserInformation is a public API surface class move it back to main package.

File: sdk/appcenter/src/test/java/com/microsoft/appcenter/utils/context/AuthTokenContextTest.java [hunk 11]
```diff
@@ -118,8 +121,8 @@ public void setAuthTokenTest() {
         mAuthTokenContext.setAuthToken(null, null, null);
 
         /* Verify that listener is called on empty token. */
-        verify(mockListener, times(1)).onNewAuthToken(isNull(String.class));
-        verify(mockListener, times(1)).onNewUser(isNull(String.class));
+         verify(mockListener).onNewAuthToken(isNull(String.class));
```
Comment (guperrot) [inline sdk/appcenter/src/test/java/com/microsoft/appcenter/utils/context/AuthTokenContextTest.java]: Code is not formatted.
I have to insist on using Android Studio commit dialog as instructions are not followed otherwise, this is not the first time.

File: sdk/appcenter/src/test/java/com/microsoft/appcenter/utils/context/AuthTokenContextTest.java [hunk 12]
```diff
@@ -118,8 +121,8 @@ public void setAuthTokenTest() {
...
+        verify(mockListener).onNewUser(isNull(UserInformation.class));
```
Comment (guperrot) [inline sdk/appcenter/src/test/java/com/microsoft/appcenter/utils/context/AuthTokenContextTest.java]: Comment was not addressed: we need better than a null check, we need the test to pass a long homeAccountId value and verify the UserInformation contains the short accountId.

File: sdk/appcenter-identity/src/main/java/com/microsoft/appcenter/identity/SignInResult.java (hunk 9 above)
Comment (guperrot) [reply sdk/appcenter-identity/src/main/java/com/microsoft/appcenter/identity/SignInResult.java]: Will be fixed by rollbacking https://github.com/Microsoft/AppCenter-SDK-Android/pull/1047#discussion_r273196308

File: sdk/appcenter-identity/src/main/java/com/microsoft/appcenter/identity/Identity.java (hunk 8 above)
Comment (guperrot) [reply sdk/appcenter-identity/src/main/java/com/microsoft/appcenter/identity/Identity.java]: Will be fixed by rollbacking https://github.com/Microsoft/AppCenter-SDK-Android/pull/1047#discussion_r273196308

File: sdk/appcenter-identity/src/test/java/com/microsoft/appcenter/identity/IdentityTest.java [hunk 13]
```diff
@@ -1140,7 +1140,7 @@ public void testRemoveAccountWhenClientAppIsNull() throws Exception {
         when(mAuthTokenContext.getHomeAccountId()).thenReturn(UUIDUtils.randomUUID().toString());
         when(mAuthTokenContext.getAuthToken()).thenReturn(UUIDUtils.randomUUID().toString());
         Identity.signOut();
-        verify(publicClientApplication, never()).getAccounts(any(PublicClientApplication.AccountsLoadedListener.class));
+        verify(publicClientApplication, never()).getAccounts(any(PublicClientApplication.AccountsLoadedCallback.class));
```
Comment (guperrot) [inline sdk/appcenter-identity/src/test/java/com/microsoft/appcenter/identity/IdentityTest.java]: revert those changes, we don't even use this method in code, it does not make sense to verify that, there should be a separate PR to improve test logic here.

Comment (guperrot) [reply sdk/appcenter-identity/src/test/java/com/microsoft/appcenter/identity/IdentityTest.java]: You can simply revert your changes and accept develop changes now for this file.

File: sdk/appcenter/src/test/java/com/microsoft/appcenter/utils/context/AuthTokenContextTest.java (hunk 11 above)
Comment (jinquan2019) [reply sdk/appcenter/src/test/java/com/microsoft/appcenter/utils/context/AuthTokenContextTest.java]: This is changed by @jwallra. I have fixed it.

File: sdk/appcenter/src/test/java/com/microsoft/appcenter/utils/context/AuthTokenContextTest.java (hunk 12 above)
Comment (jinquan2019) [reply sdk/appcenter/src/test/java/com/microsoft/appcenter/utils/context/AuthTokenContextTest.java]: Updated

File: sdk/appcenter-identity/src/test/java/com/microsoft/appcenter/identity/IdentityTest.java (hunk 13 above)
Comment (jinquan2019) [reply sdk/appcenter-identity/src/test/java/com/microsoft/appcenter/identity/IdentityTest.java]: OK, I have merged it from develop branch

File: sdk/appcenter-identity/src/main/java/com/microsoft/appcenter/identity/Identity.java [hunk 14]
```diff
@@ -16,6 +16,7 @@
 
 import com.microsoft.appcenter.AbstractAppCenterService;
 import com.microsoft.appcenter.AppCenter;
+import com.microsoft.appcenter.UserInformation;
```
Comment (guperrot) [inline sdk/appcenter-identity/src/main/java/com/microsoft/appcenter/identity/Identity.java]: Maybe I didn't explain correctly in previous comments.
Just revert this change.

File: sdk/appcenter-identity/src/main/java/com/microsoft/appcenter/identity/SignInResult.java [hunk 15]
```diff
@@ -5,6 +5,8 @@
 
 package com.microsoft.appcenter.identity;
 
+import com.microsoft.appcenter.UserInformation;
```
Comment (guperrot) [inline sdk/appcenter-identity/src/main/java/com/microsoft/appcenter/identity/SignInResult.java]: Maybe I didn't explain correctly in previous comments.
Just revert this change.

File: sdk/appcenter/src/main/java/com/microsoft/appcenter/UserInformation.java [hunk 16]
```diff
@@ -3,7 +3,7 @@
  * Licensed under the MIT License.
  */
 
-package com.microsoft.appcenter.identity;
+package com.microsoft.appcenter;
```
Comment (guperrot) [inline sdk/appcenter/src/main/java/com/microsoft/appcenter/UserInformation.java]: Maybe I didn't explain correctly in previous comments.
Just revert this file, it needs to stay `package com.microsoft.appcenter.identity;` like before.

File: sdk/appcenter/src/test/java/com/microsoft/appcenter/utils/context/AuthTokenContextTest.java [hunk 17]
```diff
@@ -97,19 +99,23 @@ public void setAuthTokenTest() {
...
         /* Verify the value is stored. */
         verifyStatic(times(2));
         SharedPreferencesManager.putString(eq(PREFERENCE_KEY_TOKEN_HISTORY), anyString());
 
         /* Verify that listener is called. */
         verify(mockListener, times(2)).onNewAuthToken(notNull(String.class));
-        verify(mockListener, times(1)).onNewUser(notNull(String.class));
+        ArgumentCaptor<UserInformation> captorArg = ArgumentCaptor.forClass(UserInformation.class);
+        verify(mockListener).onNewUser(captorArg.capture());
+        assertEquals(accountId, captorArg.getValue().getAccountId());
```
Comment (guperrot) [inline sdk/appcenter/src/test/java/com/microsoft/appcenter/utils/context/AuthTokenContextTest.java]: Ok but `assertNotNull(captorArg.getValue())` should be done before this line also.

File: sdk/appcenter/src/main/java/com/microsoft/appcenter/UserInformation.java (hunk 16 above)
Comment (jinquan2019) [reply sdk/appcenter/src/main/java/com/microsoft/appcenter/UserInformation.java]: But we can't import the package com.microsoft.appcenter.identity in the appcenter core package, because identity package is dependent on it.

Comment (guperrot) [reply sdk/appcenter/src/main/java/com/microsoft/appcenter/UserInformation.java]: Oh it's in core now ok. Then root of the core is fine.
